File: app/src/main/java/com/futsch1/medtimer/overview/ManualDose.java
package com.futsch1.medtimer.overview;

import android.app.AlertDialog;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;

import com.futsch1.medtimer.R;
import com.futsch1.medtimer.database.Medicine;
import com.futsch1.medtimer.database.MedicineRepository;
import com.futsch1.medtimer.database.MedicineWithReminders;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.database.ReminderEvent;
import com.futsch1.medtimer.helpers.DialogHelper;
import com.futsch1.medtimer.helpers.ReminderHelperKt;
import com.futsch1.medtimer.helpers.TimeHelper;

import java.time.Instant;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class ManualDose {

    private final Context context;
    private final MedicineRepository medicineRepository;
    private final FragmentActivity activity;
    private final SharedPreferences sharedPreferences;

    public ManualDose(Context context, MedicineRepository medicineRepository, FragmentActivity activity) {
        this.context = context;
        this.medicineRepository = medicineRepository;
        this.activity = activity;
        this.sharedPreferences = context.getSharedPreferences("medtimer.data", Context.MODE_PRIVATE);
    }

    public void logManualDose() {
        List<MedicineWithReminders> medicines = medicineRepository.getMedicines();
        List<ManualDoseEntry> entries = getManualDoseEntries(medicines);

        // But run the actual dialog on the UI thread again
        activity.runOnUiThread(() ->
                new AlertDialog.Builder(context)
                        .setItems(entries.stream().map(e -> e.name).toArray(String[]::new), (dialog, which) ->
                                startLogProcess(entries.get(which)))
                        .setTitle(R.string.tab_medicine)
                        .show());
    }

    @NonNull
    private List<ManualDoseEntry> getManualDoseEntries(List<MedicineWithReminders> medicines) {
        String lastCustomDose = getLastCustomDose();
        List<ManualDoseEntry> entries = new ArrayList<>();
        entries.add(new ManualDoseEntry(context.getString(R.string.custom)));
        if (!lastCustomDose.isBlank()) {
            entries.add(new ManualDoseEntry(lastCustomDose));
        }
        for (MedicineWithReminders medicine : medicines) {
            entries.add(new ManualDoseEntry(medicine.medicine, null));
            addInactiveReminders(medicine, entries);
        }
        return entries;
    }

    private void startLogProcess(ManualDoseEntry entry) {
        ReminderEvent reminderEvent = new ReminderEvent();
        // Manual dose is not assigned to an existing reminder
        reminderEvent.reminderId = -1;
        reminderEvent.status = ReminderEvent.ReminderStatus.TAKEN;
        reminderEvent.medicineName = entry.name;
        reminderEvent.color = entry.color;
        reminderEvent.useColor = entry.useColor;
        reminderEvent.iconId = entry.iconId;
        if (reminderEvent.medicineName.equals(context.getString(R.string.custom))) {
            DialogHelper.showTextInputDialog(context, R.string.log_additional_dose, R.string.medicine_name, name -> {
                setLastCustomDose(name);
                reminderEvent.medicineName = name;
                getAmountAndContinue(reminderEvent);
            });
        } else {
            if (entry.amount == null) {
                getAmountAndContinue(reminderEvent);
            } else {
                reminderEvent.amount = entry.amount;
                getTimeAndLog(reminderEvent);
            }
        }
    }

    private String getLastCustomDose() {
        return sharedPreferences.getString("lastCustomDose", "");
    }

    private static void addInactiveReminders(MedicineWithReminders medicine, List<ManualDoseEntry> entries) {
        for (Reminder reminder : medicine.reminders) {
            if (!ReminderHelperKt.isReminderActive(reminder)) {
                entries.add(new ManualDoseEntry(medicine.medicine, reminder.amount));
            }
        }
    }

    private void getAmountAndContinue(ReminderEvent reminderEvent) {
        DialogHelper.showTextInputDialog(context, R.string.log_additional_dose, R.string.dosage, amount -> {
            reminderEvent.amount = amount;
            getTimeAndLog(reminderEvent);
        });
    }

    private void getTimeAndLog(ReminderEvent reminderEvent) {
        LocalDateTime localDateTime = LocalDateTime.now();
        TimeHelper.TimePickerWrapper timePicker = new TimeHelper.TimePickerWrapper(activity);
        timePicker.show(localDateTime.getHour(), localDateTime.getMinute(), minutes -> {
            reminderEvent.remindedTimestamp = TimeHelper.instantFromTodayMinutes(minutes).getEpochSecond();
            reminderEvent.processedTimestamp = Instant.now().getEpochSecond();

            medicineRepository.insertReminderEvent(reminderEvent);
        });
    }

    private void setLastCustomDose(String lastCustomDose) {
        sharedPreferences.edit().putString("lastCustomDose", lastCustomDose).apply();
    }

    private static class ManualDoseEntry {
        public final String name;
        public final int color;
        public final boolean useColor;
        public final String amount;
        public final int iconId;

        public ManualDoseEntry(String name) {
            this.name = name;
            this.color = 0;
            this.useColor = false;
            this.amount = null;
            this.iconId = 0;
        }

        public ManualDoseEntry(Medicine medicine, String amount) {
            if (amount != null) {
                this.name = medicine.name + " (" + amount + ")";
            } else {
                this.name = medicine.name;
            }
            this.color = medicine.color;
            this.useColor = medicine.useColor;
            this.amount = amount;
            this.iconId = medicine.iconId;
        }
    }
}

File: app/src/main/java/com/futsch1/medtimer/overview/ManualDose.kt
package com.futsch1.medtimer.overview

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import androidx.core.content.edit
import androidx.fragment.app.FragmentActivity
import com.futsch1.medtimer.R
import com.futsch1.medtimer.database.FullMedicine
import com.futsch1.medtimer.database.MedicineRepository
import com.futsch1.medtimer.database.ReminderEvent
import com.futsch1.medtimer.helpers.DialogHelper
import com.futsch1.medtimer.helpers.TimeHelper
import com.futsch1.medtimer.helpers.TimeHelper.TimePickerWrapper
import com.futsch1.medtimer.helpers.isReminderActive
import com.futsch1.medtimer.reminders.ReminderProcessor
import java.time.LocalDateTime
import java.util.stream.Collectors

class ManualDose(
    private val context: Context,
    private val medicineRepository: MedicineRepository,
    private val activity: FragmentActivity
) {
    @Suppress("kotlin:S6291") // Preferences do not contain sensitive date
    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences("medtimer.data", Context.MODE_PRIVATE)

    fun logManualDose() {
        val medicines = medicineRepository.medicines
        val entries = getManualDoseEntries(medicines)
        val adapter = ManualDoseListEntryAdapter(context, R.layout.manual_dose_list_entry, entries)

        // But run the actual dialog on the UI thread again
        activity.runOnUiThread {
            AlertDialog.Builder(context)
                .setAdapter(adapter) { _: DialogInterface?, which: Int ->
                    startLogProcess(entries[which])
                }
                .setTitle(R.string.log_additional_dose)
                .show()
        }
    }

    private fun getManualDoseEntries(medicines: List<FullMedicine>): List<ManualDoseEntry> {
        val entries: MutableList<ManualDoseEntry> = ArrayList()
        entries.add(ManualDoseEntry(context.getString(R.string.custom)))
        addCustomDoses(entries)
        for (medicine in medicines) {
            val entry = ManualDoseEntry(medicine, null)
            entries.add(entry)
            addInactiveReminders(medicine, entries)
        }
        return entries
    }

    private fun addCustomDoses(
        entries: MutableList<ManualDoseEntry>
    ) {
        val lastCustomDose = lastCustomDose
        if (lastCustomDose.first != null && lastCustomDose.first!!.isNotBlank()) {
            entries.add(ManualDoseEntry(lastCustomDose.first!!))
            if (lastCustomDose.second != null && lastCustomDose.second!!.isNotBlank()) {
                entries.add(ManualDoseEntry(lastCustomDose.first!!, lastCustomDose.second))
            }
        }
    }

    private fun startLogProcess(entry: ManualDoseEntry) {
        val reminderEvent = ReminderEvent()
        // Manual dose is not assigned to an existing reminder
        reminderEvent.reminderId = -1
        reminderEvent.status = ReminderEvent.ReminderStatus.TAKEN
        reminderEvent.medicineName = entry.name
        reminderEvent.color = entry.color
        reminderEvent.useColor = entry.useColor
        reminderEvent.iconId = entry.iconId
        reminderEvent.tags = entry.tags
        if (reminderEvent.medicineName == context.getString(R.string.custom)) {
            DialogHelper(context).title(R.string.log_additional_dose).hint(R.string.medicine_name)
                .textSink { name: String? ->
                    reminderEvent.medicineName = name
                    entry.baseName = name!!
                    getAmountAndContinue(reminderEvent, entry)
                }.show()
        } else {
            if (entry.amount == null || entry.medicineId == -1) {
                getAmountAndContinue(reminderEvent, entry)
            } else {
                getTimeAndLog(reminderEvent, entry.medicineId)
            }
        }
    }

    private var lastCustomDose: Pair<String?, String?>
        get() {
            val name = sharedPreferences.getString("lastCustomDose", "")
            val amount = sharedPreferences.getString("lastCustomDoseAmount", "")
            return Pair(name, amount)
        }
        set(lastCustomDose) {
            sharedPreferences.edit { putString("lastCustomDose", lastCustomDose.first); putString("lastCustomDoseAmount", lastCustomDose.second) }
        }

    private fun getAmountAndContinue(reminderEvent: ReminderEvent, entry: ManualDoseEntry) {
        var dialog = DialogHelper(context).title(R.string.log_additional_dose).hint(R.string.dosage)
            .textSink { amount: String? ->
                reminderEvent.amount = amount
                if (entry.medicineId == -1) {
                    lastCustomDose = Pair(entry.baseName, amount)
                }
                getTimeAndLog(reminderEvent, entry.medicineId)
            }
        if (entry.amount != null && !entry.amount.isBlank()) {
            dialog = dialog.initialText(entry.amount)
        }
        dialog.show()
    }

    private fun getTimeAndLog(reminderEvent: ReminderEvent, medicineId: Int) {
        val localDateTime = LocalDateTime.now()
        val timePicker = TimePickerWrapper(activity)
        timePicker.show(localDateTime.hour, localDateTime.minute) { minutes: Int ->
            reminderEvent.remindedTimestamp =
                TimeHelper.instantFromTodayMinutes(minutes).epochSecond
            reminderEvent.processedTimestamp = reminderEvent.remindedTimestamp
            medicineRepository.insertReminderEvent(reminderEvent)
        }
        if (medicineId != -1) {
            ReminderProcessor.requestStockHandling(context, reminderEvent.amount!!, medicineId)
        }
    }

    class ManualDoseEntry {
        var baseName: String
        lateinit var name: String
        val color: Int
        val useColor: Boolean
        val amount: String?
        val iconId: Int
        val medicineId: Int
        val tags: List<String>

        constructor(name: String, amount: String? = null) {
            this.baseName = name
            this.color = 0
            this.useColor = false
            this.amount = amount
            this.iconId = 0
            this.medicineId = -1
            this.tags = ArrayList()
            amendName()
        }

        constructor(medicine: FullMedicine, amount: String?) {
            this.baseName = medicine.medicine.name
            this.color = medicine.medicine.color
            this.useColor = medicine.medicine.useColor
            this.amount = amount
            this.iconId = medicine.medicine.iconId
            this.medicineId = medicine.medicine.medicineId
            this.tags = medicine.tags.stream().map { t -> t.name }.collect(Collectors.toList())
            amendName()
        }

        private fun amendName() {
            if (amount != null) {
                this.name = this.baseName + " (" + amount + ")"
            } else {
                this.name = this.baseName
            }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other == null || javaClass != other.javaClass) return false
            val that = other as ManualDoseEntry
            return !(name != that.name || amount != that.amount)
        }

        override fun hashCode(): Int {
            return super.hashCode()
        }
    }

    companion object {
        private fun addInactiveReminders(
            medicine: FullMedicine,
            entries: MutableList<ManualDoseEntry>
        ) {
            for (reminder in medicine.reminders) {
                val entry = ManualDoseEntry(medicine, reminder.amount)
                if (!isReminderActive(reminder) && !entries.contains(entry)) {
                    entries.add(entry)
                }
            }
        }
    }
}
